src/metabase/upload.clj

```diff
@@ -612,7 +620,12 @@
           normed-name->field (m/index-by (comp normalize-column-name :name)
                                          (t2/select :model/Field :table_id (:id table) :active true))
           normed-header      (map normalize-column-name header)
-          create-auto-pk?    (not (contains? normed-name->field auto-pk-column-name))
+          ;; Create auto-pk columns for drivers that supported uploads before auto-pk columns
+          ;; were introduced by metabase#36249. Otherwise we can assume that the table was created
+          ;; with an auto-pk column, or it has been deliberately removed.
+          create-auto-pk?    (and
+                              (contains? #{:mysql :postgres :h2} (:engine database))
+                              (not (contains? normed-name->field auto-pk-column-name)))
```

Comment:
If it's already an established idiom, I'm happy. 

Thanks for the reference, I see that code even made non-local assumptions that the value was a concrete map, and in this case at least the variable is defined locally and it's even less ambiguous.

I've been meaning to read Elements of Clojure, so thanks for the excuse. 

> The naming is ambiguous to me

Sorry I feel like I'm missing something here - if it's ambiguous, I would to only assume the lowest common denominator, that it is an `IFn`. I guess a reasonable heuristic to follow that if it's an argument or local binding then it's "data" and hence a map, and global references can be looked up quickly to check.

Reply:
By this

> The naming is ambiguous to me, x->y is either a map or a function

I mean x->y is either an associative map (implementing `clojure.lang.Associative`) or a function that does not. And I mean you cannot tell from the name alone.

Reply:
Gotcha on the ambiguity, that's what I thought you meant. 

My original point was that if something can be A or B, and A is a kind of B, then I would think you should code against the weaker assumption that it is some kind of B. In this case A is `Associative` and B is `IFn`. We can code to that weaker assumption by using `(nil? (a->b k))` instead of `(not (contains? a->b k))` - it's even shorter, although slightly less semantically clear.

I'm happy to move on though, will just bear in mind there's always burden on the reader to be careful with making the stronger "it's data" assumption.

Reply:
Discovering points of divergence is how I learn so I'm happy to keep discussing this until we understand each other.

In this case you are correct in the sense that the name does imply that it is A or B, and if we only had that information on x->y I would agree, we should code against the weaker assumption. But we don't only have the name, we have the ability to look up the actual interface of x->y, from which we know that it is a map.

I think you are actually arguing that it is better to have names that disambiguate types and to that I agree that is a benefit of the `b-by-a` convention. OTOH this naming convention has the benefits I listed above, and I think its downsides are limited if used as a local variable name.

Reply:
on this

> We can code to that weaker assumption by using `(nil? (a->b k))` instead of `(not (contains? a->b k))`

I'm not sure I would prefer that because if `a->b` was a function, the name alone doesn't imply that it returns `nil` for an arg that isn't in its domain. So it doesn't really fit the weaker assumption implied by `a->b` and actually puts quite strong assumptions of the implementation of the hypothetical function `a->b`. I think I prefer `(contains? a->b k)` because at least it's making the assumption clear that `a->b` implements Associative and those semantics are well understood. And a function could return nil for many reasons.

Reply:
It took me we a while, but I think the penny dropped at how we were missing each other.

I believe you're saying that we can apply duck typing. We know that this variable is associative *because* there is code that calls `contains?` on it.

My concern was about checking that it's safe to add that call in the first place, and the long term safety of the assumption - for example a far away definition could get refactored from a map to a function. I guess it's on the reviewer to check and the tests to give confidence, and that's fine.

A long thread but valuable for to remember that kind "duck" thinking - seeing every method call as type information for the variable they're being passed. It's been a very long time since I worked in a dynamic language and this was a useful insight.